From b8c26fd86281f3b0883c31bd2d0cb467a12b860d Mon Sep 17 00:00:00 2001
From: Jerome Soumagne <jerome.soumagne@intel.com>
Date: Thu, 7 Dec 2023 11:34:42 -0600
Subject: [PATCH] NA OFI: remove excessive MR count warning message

---
 src/na/na_ofi.c | 8 --------
 1 file changed, 8 deletions(-)

diff --git a/src/na/na_ofi.c b/src/na/na_ofi.c
index 2745d8ae..57945d8f 100644
--- a/src/na/na_ofi.c
+++ b/src/na/na_ofi.c
@@ -8163,14 +8163,6 @@ na_ofi_mem_register(na_class_t *na_class, na_mem_handle_t *mem_handle,
     na_return_t ret;
     int rc;
 
-    /* Just throw a warning if we start exceeding the optimal number of
-     * MRs for that domain */
-    NA_CHECK_SUBSYS_WARNING(mem,
-        fi_info->domain_attr->mr_cnt > 0 &&
-            !((size_t) mr_cnt < fi_info->domain_attr->mr_cnt),
-        "Exceeding domain's optimal MR count (%" PRId32 " >= %zu)", mr_cnt,
-        fi_info->domain_attr->mr_cnt);
-
     /* Set access mode */
     switch (na_ofi_mem_handle->desc.info.flags) {
         case NA_MEM_READ_ONLY:
