diff --git a/src/na/na_ucx.c b/src/na/na_ucx.c
index f8fbb23..ea13af1 100644
--- a/src/na/na_ucx.c
+++ b/src/na/na_ucx.c
@@ -1,6 +1,7 @@
 /**
  * Copyright (c) 2013-2022 UChicago Argonne, LLC and The HDF Group.
- * Copyright (c) 2022-2023 Intel Corporation.
+ * Copyright (c) 2022-2024 Intel Corporation.
+ * Copyright (c) 2024-2025 Hewlett Packard Enterprise Development LP.
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -80,11 +81,11 @@
 #define NA_UCX_OP_RESET(_op, _context, _cb_type, _cb, _arg, _addr)             \
     do {                                                                       \
         _op->context = _context;                                               \
-        _op->completion_data = (struct na_cb_completion_data){                 \
+        _op->completion_data = (struct na_cb_completion_data) {                \
             .callback_info =                                                   \
-                (struct na_cb_info){                                           \
+                (struct na_cb_info) {                                          \
                     .info.recv_unexpected =                                    \
-                        (struct na_cb_info_recv_unexpected){                   \
+                        (struct na_cb_info_recv_unexpected) {                  \
                             .actual_buf_size = 0, .source = NULL, .tag = 0},   \
                     .arg = _arg,                                               \
                     .type = _cb_type,                                          \
@@ -128,16 +129,18 @@
 
 /* Address */
 struct na_ucx_addr {
-    STAILQ_ENTRY(na_ucx_addr) entry;   /* Entry in addr pool */
-    struct sockaddr_storage ss_addr;   /* Sock addr */
-    ucs_sock_addr_t addr_key;          /* Address key */
-    struct na_ucx_class *na_ucx_class; /* NA UCX class */
-    ucp_address_t *worker_addr;        /* Worker addr */
-    size_t worker_addr_len;            /* Worker addr len */
-    bool worker_addr_alloc;            /* Worker addr was allocated by us */
-    ucp_ep_h ucp_ep;                   /* Currently only one EP per address */
-    hg_atomic_int32_t refcount;        /* Reference counter */
-    hg_atomic_int32_t status;          /* Connection state */
+    STAILQ_ENTRY(na_ucx_addr) entry;          /* Entry in addr pool */
+    LIST_ENTRY(na_ucx_addr) close_list_entry; /* Entry in close list */
+    struct sockaddr_storage ss_addr;          /* Sock addr */
+    ucs_sock_addr_t addr_key;                 /* Address key */
+    struct na_ucx_class *na_ucx_class;        /* NA UCX class */
+    ucp_address_t *worker_addr;               /* Worker addr */
+    size_t worker_addr_len;                   /* Worker addr len */
+    bool worker_addr_alloc;     /* Worker addr was allocated by us */
+    ucp_ep_h ucp_ep;            /* Currently only one EP per address */
+    void *close_request;        /* Close request */
+    hg_atomic_int32_t refcount; /* Reference counter */
+    hg_atomic_int32_t status;   /* Connection state */
 };
 
 /* Map (used to cache addresses) */
@@ -216,6 +219,12 @@ struct na_ucx_addr_pool {
     hg_thread_spin_t lock;
 };
 
+/* Addr list */
+struct na_ucx_addr_list {
+    LIST_HEAD(, na_ucx_addr) list;
+    hg_thread_spin_t lock;
+};
+
 /* Unexpected msg info */
 struct na_ucx_unexpected_info {
     STAILQ_ENTRY(na_ucx_unexpected_info) entry;
@@ -245,6 +254,7 @@ struct na_ucx_class {
     struct na_ucx_map addr_map;                 /* Address map */
     struct na_ucx_op_queue unexpected_op_queue; /* Unexpected op queue */
     struct na_ucx_addr_pool addr_pool;          /* Addr pool */
+    struct na_ucx_addr_list addr_close_list;    /* List of addresses to close */
     ucp_context_h ucp_context;                  /* UCP context */
     ucp_worker_h ucp_worker;                    /* Shared UCP worker */
     ucp_listener_h ucp_listener;   /* Listener handle if listening */
@@ -445,7 +455,7 @@ na_ucp_ep_error_cb(void *arg, ucp_ep_h ep, ucs_status_t status);
  * Close endpoint.
  */
 static void
-na_ucp_ep_close(ucp_ep_h ep);
+na_ucp_ep_close(ucp_ep_h ep, bool force, void **request_p);
 
 #ifndef NA_UCX_HAS_MEM_POOL
 /**
@@ -614,7 +624,7 @@ na_ucx_addr_map_update(struct na_ucx_class *na_ucx_class,
  */
 static na_return_t
 na_ucx_addr_map_remove(
-    struct na_ucx_map *na_ucx_map, ucs_sock_addr_t *addr_key);
+    struct na_ucx_map *na_ucx_map, struct na_ucx_addr *remove_addr);
 
 /**
  * Hash connection ID.
@@ -658,7 +668,12 @@ na_ucx_addr_pool_get(struct na_ucx_class *na_ucx_class);
  * Release address without destroying it.
  */
 static void
-na_ucx_addr_release(struct na_ucx_addr *na_ucx_addr);
+na_ucx_addr_release(struct na_ucx_addr *na_ucx_addr, bool force,
+    void (*free_cb)(struct na_ucx_addr *));
+
+/* Check close list and free entries. */
+static unsigned int
+na_ucx_addr_close_list_progress(struct na_ucx_class *na_ucx_class);
 
 /**
  * Reset address.
@@ -673,6 +688,10 @@ static na_return_t
 na_ucx_addr_create(struct na_ucx_class *na_ucx_class, ucs_sock_addr_t *addr_key,
     struct na_ucx_addr **na_ucx_addr_p);
 
+/* Free address once EP is closed. */
+static void
+na_ucx_addr_free_cb(struct na_ucx_addr *na_ucx_addr);
+
 /**
  * Increment ref count.
  */
@@ -798,8 +817,8 @@ na_ucx_addr_serialize(
 
 /* addr_deserialize */
 static na_return_t
-na_ucx_addr_deserialize(
-    na_class_t *na_class, na_addr_t **addr_p, const void *buf, size_t buf_size);
+na_ucx_addr_deserialize(na_class_t *na_class, na_addr_t **addr_p,
+    const void *buf, size_t buf_size);
 
 /* msg_get_max_unexpected_size */
 static NA_INLINE size_t
@@ -1107,12 +1126,12 @@ static enum hg_log_level
 na_ucs_log_level_to_hg(ucs_log_level_t level)
 {
     switch (level) {
-        case UCS_LOG_LEVEL_FATAL:
         case UCS_LOG_LEVEL_ERROR:
             return HG_LOG_LEVEL_ERROR;
         case UCS_LOG_LEVEL_WARN:
             return HG_LOG_LEVEL_WARNING;
         case UCS_LOG_LEVEL_DIAG:
+            return HG_LOG_LEVEL_MIN_DEBUG;
         case UCS_LOG_LEVEL_INFO:
         case UCS_LOG_LEVEL_DEBUG:
         case UCS_LOG_LEVEL_TRACE:
@@ -1139,7 +1158,7 @@ na_ucs_log_level_to_string(enum hg_log_level level)
         case HG_LOG_LEVEL_WARNING:
             return "warn";
         case HG_LOG_LEVEL_MIN_DEBUG:
-            return "trace";
+            return "diag";
         case HG_LOG_LEVEL_DEBUG:
             return "debug";
         case HG_LOG_LEVEL_NONE:
@@ -1623,7 +1642,7 @@ na_ucp_listener_create(ucp_worker_h worker, const struct sockaddr *addr,
         .field_mask = UCP_LISTENER_PARAM_FIELD_SOCK_ADDR |
                       UCP_LISTENER_PARAM_FIELD_CONN_HANDLER,
         .sockaddr = (ucs_sock_addr_t){.addr = addr, .addrlen = addrlen},
-        .conn_handler = (ucp_listener_conn_handler_t){
+        .conn_handler = (ucp_listener_conn_handler_t) {
             .cb = na_ucp_listener_conn_cb, .arg = listener_arg}};
     ucp_listener_attr_t listener_attrs = {
         .field_mask = UCP_LISTENER_ATTR_FIELD_SOCKADDR};
@@ -1687,12 +1706,16 @@ na_ucp_listener_conn_cb(ucp_conn_request_h conn_request, void *arg)
         error, "conn attributes contain no client addr");
 
     /* Lookup address from table */
-    addr_key = (ucs_sock_addr_t){
+    addr_key = (ucs_sock_addr_t) {
         .addr = (const struct sockaddr *) &conn_request_attrs.client_address,
         .addrlen = sizeof(conn_request_attrs.client_address)};
     na_ucx_addr = na_ucx_addr_map_lookup(&na_ucx_class->addr_map, &addr_key);
-    NA_CHECK_SUBSYS_ERROR_NORET(addr, na_ucx_addr != NULL, error,
-        "An entry is already present for this address");
+
+    if (na_ucx_addr != NULL) {
+        NA_LOG_SUBSYS_WARNING(addr,
+            "An entry is already present for this address");
+        na_ucx_addr_map_remove(&na_ucx_class->addr_map, na_ucx_addr);
+    }
 
     /* Insert new entry and create new address */
     na_ret = na_ucx_addr_map_insert(na_ucx_class, &na_ucx_class->addr_map,
@@ -1727,7 +1750,7 @@ na_ucp_connect(ucp_worker_h worker, const struct sockaddr *src_addr,
     ucp_ep_params_t ep_params = {
         .field_mask = UCP_EP_PARAM_FIELD_FLAGS | UCP_EP_PARAM_FIELD_SOCK_ADDR,
         .flags = UCP_EP_PARAMS_FLAGS_CLIENT_SERVER,
-        .sockaddr = (ucs_sock_addr_t){.addr = dst_addr, .addrlen = addrlen},
+        .sockaddr = (ucs_sock_addr_t) {.addr = dst_addr, .addrlen = addrlen},
         .conn_request = NULL};
     struct sockaddr_storage src_ss_addr;
     na_return_t ret;
@@ -1938,13 +1961,25 @@ na_ucp_ep_error_cb(
 
 /*---------------------------------------------------------------------------*/
 static void
-na_ucp_ep_close(ucp_ep_h ep)
+na_ucp_ep_close(ucp_ep_h ep, bool force, void **request_p)
 {
-    ucs_status_ptr_t status_ptr = ucp_ep_close_nb(ep, UCP_EP_CLOSE_MODE_FORCE);
-    NA_CHECK_SUBSYS_ERROR_DONE(addr,
-        status_ptr != NULL && UCS_PTR_IS_ERR(status_ptr),
-        "ucp_ep_close_nb() failed (%s)",
-        ucs_status_string(UCS_PTR_STATUS(status_ptr)));
+    const ucp_request_param_t close_params = {
+        .op_attr_mask = UCP_OP_ATTR_FIELD_FLAGS,
+        .flags = force ? UCP_EP_CLOSE_FLAG_FORCE : 0};
+    ucs_status_ptr_t status_ptr;
+
+    status_ptr = ucp_ep_close_nbx(ep, &close_params);
+    if (status_ptr == NULL) /* Check for immediate completion */
+        NA_LOG_SUBSYS_DEBUG(addr, "ucp_ep_close_nbx() completed immediately");
+    else if (UCS_PTR_IS_ERR(status_ptr)) {
+        /* Keep as debug as we can't do anything anyway */
+        NA_LOG_SUBSYS_DEBUG(addr, "ucp_ep_close_nbx() failed (%s)",
+            ucs_status_string(UCS_PTR_STATUS(status_ptr)));
+    } else {
+        NA_LOG_SUBSYS_DEBUG(
+            addr, "ucp_ep_close_nbx() did not immediately complete");
+        *request_p = status_ptr;
+    }
 }
 
 /*---------------------------------------------------------------------------*/
@@ -2041,7 +2076,7 @@ na_ucp_am_recv(
 
         /* Fill unexpected info */
         na_ucx_op_id->completion_data.callback_info.info.recv_unexpected =
-            (struct na_cb_info_recv_unexpected){
+            (struct na_cb_info_recv_unexpected) {
                 .tag = (na_tag_t) na_ucx_unexpected_info->tag,
                 .actual_buf_size = (size_t) na_ucx_unexpected_info->length,
                 .source = (na_addr_t *) na_ucx_unexpected_info->na_ucx_addr};
@@ -2102,7 +2137,7 @@ na_ucp_am_recv_cb(void *arg, const void *header, size_t header_length,
     if (likely(na_ucx_op_id)) {
         /* Fill info */
         na_ucx_op_id->completion_data.callback_info.info.recv_unexpected =
-            (struct na_cb_info_recv_unexpected){.tag = (na_tag_t) tag,
+            (struct na_cb_info_recv_unexpected) {.tag = (na_tag_t) tag,
                 .actual_buf_size = (size_t) length,
                 .source = (na_addr_t *) source_addr};
         na_ucx_addr_ref_incr(source_addr);
@@ -2415,6 +2450,12 @@ na_ucx_class_alloc(void)
         cls, rc != HG_UTIL_SUCCESS, error, "hg_thread_spin_init() failed");
     STAILQ_INIT(&na_ucx_class->addr_pool.queue);
 
+    /* Initialize addr close list */
+    rc = hg_thread_spin_init(&na_ucx_class->addr_close_list.lock);
+    NA_CHECK_SUBSYS_ERROR_NORET(
+        cls, rc != HG_UTIL_SUCCESS, error, "hg_thread_spin_init() failed");
+    LIST_INIT(&na_ucx_class->addr_close_list.list);
+
     /* Create address map */
     na_ucx_class->addr_map.key_map =
         hg_hash_table_new(na_ucx_addr_key_hash, na_ucx_addr_key_equal);
@@ -2462,6 +2503,7 @@ na_ucx_class_free(struct na_ucx_class *na_ucx_class)
     (void) hg_thread_spin_destroy(&na_ucx_class->unexpected_op_queue.lock);
     (void) hg_thread_spin_destroy(&na_ucx_class->unexpected_msg_queue.lock);
     (void) hg_thread_spin_destroy(&na_ucx_class->addr_pool.lock);
+    (void) hg_thread_spin_destroy(&na_ucx_class->addr_close_list.lock);
 
     free(na_ucx_class->protocol_name);
     free(na_ucx_class);
@@ -2672,6 +2714,7 @@ static na_return_t
 na_ucx_addr_map_update(struct na_ucx_class *na_ucx_class,
     struct na_ucx_map *na_ucx_map, struct na_ucx_addr *na_ucx_addr)
 {
+    void *request = NULL;
     na_return_t ret = NA_SUCCESS;
     int rc;
 
@@ -2690,9 +2733,11 @@ na_ucx_addr_map_update(struct na_ucx_class *na_ucx_class,
     NA_CHECK_SUBSYS_ERROR(addr, rc != 1, unlock, ret, NA_NOENTRY,
         "hg_hash_table_remove() failed");
 
-    /* Close previous EP */
-    na_ucp_ep_close(na_ucx_addr->ucp_ep);
+    /* Close previous EP (force closure) */
+    na_ucp_ep_close(na_ucx_addr->ucp_ep, true, &request);
     na_ucx_addr->ucp_ep = NULL;
+    NA_CHECK_SUBSYS_ERROR(addr, request != NULL, unlock, ret, NA_BUSY,
+        "Could not close UCP endpoint");
 
     /* Create new endpoint */
     ret = na_ucp_connect(na_ucx_class->ucp_worker,
@@ -2725,7 +2770,8 @@ unlock:
 
 /*---------------------------------------------------------------------------*/
 static na_return_t
-na_ucx_addr_map_remove(struct na_ucx_map *na_ucx_map, ucs_sock_addr_t *addr_key)
+na_ucx_addr_map_remove(
+    struct na_ucx_map *na_ucx_map, struct na_ucx_addr *remove_addr)
 {
     struct na_ucx_addr *na_ucx_addr = NULL;
     na_return_t ret = NA_SUCCESS;
@@ -2734,13 +2780,15 @@ na_ucx_addr_map_remove(struct na_ucx_map *na_ucx_map, ucs_sock_addr_t *addr_key)
     hg_thread_rwlock_wrlock(&na_ucx_map->lock);
 
     na_ucx_addr = hg_hash_table_lookup(
-        na_ucx_map->key_map, (hg_hash_table_key_t) addr_key);
-    if (na_ucx_addr == HG_HASH_TABLE_NULL)
+        na_ucx_map->key_map, (hg_hash_table_key_t) &remove_addr->addr_key);
+
+    if (na_ucx_addr == HG_HASH_TABLE_NULL ||
+        na_ucx_addr->ucp_ep != remove_addr->ucp_ep)
         goto unlock;
 
     /* Remove addr key from primary map */
     rc = hg_hash_table_remove(
-        na_ucx_map->key_map, (hg_hash_table_key_t) addr_key);
+        na_ucx_map->key_map, (hg_hash_table_key_t) &na_ucx_addr->addr_key);
     NA_CHECK_SUBSYS_ERROR(addr, rc != 1, unlock, ret, NA_NOENTRY,
         "hg_hash_table_remove() failed");
 
@@ -2809,10 +2857,12 @@ na_ucx_addr_destroy(struct na_ucx_addr *na_ucx_addr)
 {
     NA_LOG_SUBSYS_DEBUG(addr, "Destroying address %p", (void *) na_ucx_addr);
 
-    na_ucx_addr_release(na_ucx_addr);
+    na_ucx_addr_release(na_ucx_addr, true, NULL);
     free(na_ucx_addr);
 }
 
+
+
 /*---------------------------------------------------------------------------*/
 #ifdef NA_UCX_HAS_ADDR_POOL
 static struct na_ucx_addr *
@@ -2837,22 +2887,15 @@ na_ucx_addr_pool_get(struct na_ucx_class *na_ucx_class)
 
 /*---------------------------------------------------------------------------*/
 static void
-na_ucx_addr_release(struct na_ucx_addr *na_ucx_addr)
+na_ucx_addr_release(struct na_ucx_addr *na_ucx_addr, bool force,
+    void (*free_cb)(struct na_ucx_addr *))
 {
     /* Make sure we remove from map before we close the EP */
     if (na_ucx_addr->addr_key.addr) {
         NA_UCX_PRINT_ADDR_KEY_INFO("Removing address", &na_ucx_addr->addr_key);
 
         na_ucx_addr_map_remove(
-            &na_ucx_addr->na_ucx_class->addr_map, &na_ucx_addr->addr_key);
-    }
-
-    if (na_ucx_addr->ucp_ep != NULL) {
-        /* NB. for deserialized addresses that are not "connected" addresses, do
-         * not close the EP */
-        if (na_ucx_addr->worker_addr == NULL)
-            na_ucp_ep_close(na_ucx_addr->ucp_ep);
-        na_ucx_addr->ucp_ep = NULL;
+            &na_ucx_addr->na_ucx_class->addr_map, na_ucx_addr);
     }
 
     if (na_ucx_addr->worker_addr != NULL) {
@@ -2864,9 +2907,59 @@ na_ucx_addr_release(struct na_ucx_addr *na_ucx_addr)
         na_ucx_addr->worker_addr = NULL;
         na_ucx_addr->worker_addr_len = 0;
     }
+
+    if (na_ucx_addr->ucp_ep != NULL) {
+        /* Close endpoint (if force is not passed, operations will flush) */
+        na_ucp_ep_close(
+            na_ucx_addr->ucp_ep, force, &na_ucx_addr->close_request);
+        na_ucx_addr->ucp_ep = NULL;
+
+        if (na_ucx_addr->close_request != NULL) {
+            NA_LOG_SUBSYS_DEBUG(addr, "EP close for addr %p still pending",
+                (void *) na_ucx_addr);
+            /* Keep addr in close list to progress later */
+            hg_thread_spin_lock(
+                &na_ucx_addr->na_ucx_class->addr_close_list.lock);
+            LIST_INSERT_HEAD(&na_ucx_addr->na_ucx_class->addr_close_list.list,
+                na_ucx_addr, close_list_entry);
+            hg_thread_spin_unlock(
+                &na_ucx_addr->na_ucx_class->addr_close_list.lock);
+        } else if (free_cb)
+            free_cb(na_ucx_addr);
+    }
 }
 
 /*---------------------------------------------------------------------------*/
+static unsigned int
+na_ucx_addr_close_list_progress(struct na_ucx_class *na_ucx_class)
+{
+    unsigned int count = 0;
+
+    hg_thread_spin_lock(&na_ucx_class->addr_close_list.lock);
+    if (!LIST_EMPTY(&na_ucx_class->addr_close_list.list)) {
+        struct na_ucx_addr *na_ucx_addr =
+            LIST_FIRST(&na_ucx_class->addr_close_list.list);
+        while (na_ucx_addr != NULL) {
+            struct na_ucx_addr *next = LIST_NEXT(na_ucx_addr, close_list_entry);
+            ucs_status_t status =
+                ucp_request_check_status(na_ucx_addr->close_request);
+            if (status != UCS_INPROGRESS) {
+                NA_LOG_SUBSYS_DEBUG(addr, "EP close for addr %p completed",
+                    (void *) na_ucx_addr);
+                ucp_request_free(na_ucx_addr->close_request);
+                na_ucx_addr->close_request = NULL;
+                LIST_REMOVE(na_ucx_addr, close_list_entry);
+                na_ucx_addr_free_cb(na_ucx_addr);
+                count++;
+            }
+            na_ucx_addr = next;
+        }
+    }
+    hg_thread_spin_unlock(&na_ucx_class->addr_close_list.lock);
+
+    return count;
+}
+
 static void
 na_ucx_addr_reset(struct na_ucx_addr *na_ucx_addr, ucs_sock_addr_t *addr_key)
 {
@@ -2883,7 +2976,7 @@ na_ucx_addr_reset(struct na_ucx_addr *na_ucx_addr, ucs_sock_addr_t *addr_key)
         na_ucx_addr->addr_key.addrlen = addr_key->addrlen;
     } else {
         memset(&na_ucx_addr->ss_addr, 0, sizeof(na_ucx_addr->ss_addr));
-        na_ucx_addr->addr_key = (ucs_sock_addr_t){.addr = NULL, .addrlen = 0};
+        na_ucx_addr->addr_key = (ucs_sock_addr_t) {.addr = NULL, .addrlen = 0};
     }
 }
 
@@ -2918,6 +3011,22 @@ error:
     return ret;
 }
 
+/*---------------------------------------------------------------------------*/
+static void
+na_ucx_addr_free_cb(struct na_ucx_addr *na_ucx_addr)
+{
+#ifdef NA_UCX_HAS_ADDR_POOL
+    struct na_ucx_addr_pool *addr_pool = &na_ucx_addr->na_ucx_class->addr_pool;
+
+    /* Push address back to addr pool */
+    hg_thread_spin_lock(&addr_pool->lock);
+    STAILQ_INSERT_TAIL(&addr_pool->queue, na_ucx_addr, entry);
+    hg_thread_spin_unlock(&addr_pool->lock);
+#else
+    free(na_ucx_addr);
+#endif
+}
+
 /*---------------------------------------------------------------------------*/
 static NA_INLINE void
 na_ucx_addr_ref_incr(struct na_ucx_addr *na_ucx_addr)
@@ -2937,20 +3046,8 @@ na_ucx_addr_ref_decr(struct na_ucx_addr *na_ucx_addr)
         (void *) na_ucx_addr, refcount);
 
     if (refcount == 0) {
-#ifdef NA_UCX_HAS_ADDR_POOL
-        struct na_ucx_addr_pool *addr_pool =
-            &na_ucx_addr->na_ucx_class->addr_pool;
-
         NA_LOG_SUBSYS_DEBUG(addr, "Releasing address %p", (void *) na_ucx_addr);
-        na_ucx_addr_release(na_ucx_addr);
-
-        /* Push address back to addr pool */
-        hg_thread_spin_lock(&addr_pool->lock);
-        STAILQ_INSERT_TAIL(&addr_pool->queue, na_ucx_addr, entry);
-        hg_thread_spin_unlock(&addr_pool->lock);
-#else
-        na_ucx_addr_destroy(na_ucx_addr);
-#endif
+        na_ucx_addr_release(na_ucx_addr, false, na_ucx_addr_free_cb);
     }
 }
 
@@ -3013,6 +3110,21 @@ na_ucx_rma(struct na_ucx_class NA_UNUSED *na_ucx_class, na_context_t *context,
         ret, NA_BUSY, "Attempting to use OP ID that was not completed (%s)",
         na_cb_type_to_string(na_ucx_op_id->completion_data.callback_info.type));
 
+    /* There is no need to have a fully resolved address to start an RMA.
+     * This is only necessary for two-sided communication. */
+    /* The above assumption is now in question, so the following will resolve
+     * the address if required. */
+
+    /* Check addr to ensure the EP for that addr is still valid */
+    if (!(hg_atomic_get32(&na_ucx_addr->status) & NA_UCX_ADDR_RESOLVED)) {
+        ret = na_ucx_addr_map_update(
+            na_ucx_class, &na_ucx_class->addr_map, na_ucx_addr);
+        NA_CHECK_SUBSYS_NA_ERROR(
+            addr, error, ret, "Could not update NA UCX address");
+    }
+    NA_CHECK_SUBSYS_ERROR(rma, na_ucx_addr->ucp_ep == NULL, error, ret,
+        NA_ADDRNOTAVAIL, "UCP endpoint is NULL for that address");
+
     NA_UCX_OP_RESET(na_ucx_op_id, context, cb_type, callback, arg, na_ucx_addr);
 
     na_ucx_op_id->info.rma.ucp_rma_op =
@@ -3024,9 +3136,6 @@ na_ucx_rma(struct na_ucx_class NA_UNUSED *na_ucx_class, na_context_t *context,
     na_ucx_op_id->info.rma.buf_size = length;
     na_ucx_op_id->info.rma.remote_key = NULL;
 
-    /* There is no need to have a fully resolved address to start an RMA.
-     * This is only necessary for two-sided communication. */
-
     /* TODO UCX requires the remote key to be bound to the origin, do we need a
      * new API? */
     ret = na_ucx_rma_key_resolve(na_ucx_addr->ucp_ep, remote_mem_handle,
@@ -3064,6 +3173,9 @@ na_ucx_rma_key_resolve(ucp_ep_h ep, struct na_ucx_mem_handle *na_ucx_mem_handle,
 
     hg_thread_mutex_lock(&na_ucx_mem_handle->rkey_unpack_lock);
 
+    NA_CHECK_SUBSYS_ERROR(
+        mem, ep == NULL, error, ret, NA_INVALID_ARG, "Invalid endpoint (%p)", ep);
+
     switch (hg_atomic_get32(&na_ucx_mem_handle->type)) {
         case NA_UCX_MEM_HANDLE_REMOTE_PACKED: {
             ucs_status_t status = ucp_ep_rkey_unpack(ep,
@@ -3432,6 +3544,10 @@ na_ucx_finalize(na_class_t *na_class)
     }
 #endif
 
+    NA_CHECK_SUBSYS_WARNING(cls,
+        !LIST_EMPTY(&na_ucx_class->addr_close_list.list),
+        "There are still addresses with an EP being closed");
+
     na_ucx_class_free(na_ucx_class);
     na_class->plugin_class = NULL;
 
@@ -3490,7 +3606,7 @@ na_ucx_addr_lookup(na_class_t *na_class, const char *name, na_addr_t **addr_p)
     int rc;
 
     /* Only support 'all' or same protocol */
-    NA_CHECK_SUBSYS_ERROR(fatal,
+    NA_CHECK_SUBSYS_ERROR(addr,
         strncmp(name, "all", strlen("all")) &&
             strncmp(name, na_ucx_class->protocol_name,
                 strlen(na_ucx_class->protocol_name)),
@@ -3516,7 +3632,7 @@ na_ucx_addr_lookup(na_class_t *na_class, const char *name, na_addr_t **addr_p)
         "getaddrinfo() failed (%s)", gai_strerror(rc));
 
     /* Lookup address from table */
-    addr_key = (ucs_sock_addr_t){
+    addr_key = (ucs_sock_addr_t) {
         .addr = hostname_res->ai_addr, .addrlen = hostname_res->ai_addrlen};
     na_ucx_addr = na_ucx_addr_map_lookup(&na_ucx_class->addr_map, &addr_key);
 
@@ -3668,8 +3784,8 @@ done:
 
 /*---------------------------------------------------------------------------*/
 static na_return_t
-na_ucx_addr_deserialize(
-    na_class_t *na_class, na_addr_t **addr_p, const void *buf, size_t buf_size)
+na_ucx_addr_deserialize(na_class_t *na_class, na_addr_t **addr_p,
+    const void *buf, size_t buf_size)
 {
     struct na_ucx_class *na_ucx_class = NA_UCX_CLASS(na_class);
     struct na_ucx_addr *na_ucx_addr = NULL;
@@ -3706,6 +3822,8 @@ na_ucx_addr_deserialize(
         na_ucp_ep_error_cb, na_ucx_addr, &na_ucx_addr->ucp_ep);
     NA_CHECK_SUBSYS_NA_ERROR(
         addr, error, ret, "Could not connect to remote worker");
+    //NA_LOG_SUBSYS_WARNING(addr, "Connected in deserialize, worker: %p, ep: %p",
+    //    na_ucx_class->ucp_worker, na_ucx_addr->ucp_ep);
 
     hg_atomic_or32(&na_ucx_addr->status, NA_UCX_ADDR_RESOLVED);
 
@@ -4267,11 +4385,18 @@ static NA_INLINE bool
 na_ucx_poll_try_wait(na_class_t *na_class, na_context_t NA_UNUSED *context)
 {
     struct na_ucx_class *na_ucx_class = NA_UCX_CLASS(na_class);
+    bool addr_list_empty;
     ucs_status_t status;
 
     if (na_ucx_class->no_wait)
         return false;
 
+    hg_thread_spin_lock(&na_ucx_class->addr_close_list.lock);
+    addr_list_empty = LIST_EMPTY(&na_ucx_class->addr_close_list.list);
+    hg_thread_spin_unlock(&na_ucx_class->addr_close_list.lock);
+    if (!addr_list_empty)
+        return false;
+
     status = ucp_worker_arm(na_ucx_class->ucp_worker);
     if (status == UCS_ERR_BUSY) {
         /* Events have already arrived */
@@ -4290,8 +4415,10 @@ static NA_INLINE na_return_t
 na_ucx_poll(na_class_t *na_class, na_context_t NA_UNUSED *context,
     unsigned int *count_p)
 {
-    unsigned int count =
-        ucp_worker_progress(NA_UCX_CLASS(na_class)->ucp_worker);
+    struct na_ucx_class *na_ucx_class = NA_UCX_CLASS(na_class);
+    unsigned int count = ucp_worker_progress(na_ucx_class->ucp_worker);
+    count += na_ucx_addr_close_list_progress(na_ucx_class);
+
     if (count_p != NULL)
         *count_p = count;
 
