From 320f8c68edd13ba221fa599d15658dd1a78a87e2 Mon Sep 17 00:00:00 2001
From: Jerome Soumagne <jerome.soumagne@intel.com>
Date: Thu, 4 May 2023 15:53:22 -0500
Subject: [PATCH 2/2] NA OFI: remove NA_OFI_WAIT_FD for CXI

---
 src/na/na_ofi.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/na/na_ofi.c b/src/na/na_ofi.c
index 178ca4c9..43055b37 100644
--- a/src/na/na_ofi.c
+++ b/src/na/na_ofi.c
@@ -220,7 +220,7 @@
       FI_PROGRESS_MANUAL,                                                      \
       FI_PROTO_CXI,                                                            \
       FI_MULTI_RECV,                                                           \
-      NA_OFI_WAIT_FD | NA_OFI_LOC_INFO | NA_OFI_HMEM                           \
+      NA_OFI_LOC_INFO | NA_OFI_HMEM                                            \
     )                                                                          \
     X(NA_OFI_PROV_MAX, "", "", 0, 0, 0, 0, 0, 0)
 /* clang-format on */
-- 
2.40.0

