diff --git a/src/na/na_ofi.c b/src/na/na_ofi.c
index 57059173..c6e73321 100644
--- a/src/na/na_ofi.c
+++ b/src/na/na_ofi.c
@@ -223,7 +223,7 @@
       FI_PROGRESS_MANUAL,                                                      \
       FI_PROTO_CXI,                                                            \
       FI_SOURCE | FI_MULTI_RECV,                                               \
-      NA_OFI_WAIT_FD | NA_OFI_LOC_INFO | NA_OFI_HMEM                           \
+      NA_OFI_LOC_INFO | NA_OFI_HMEM                                            \
     )                                                                          \
     X(NA_OFI_PROV_MAX, "", "", 0, 0, 0, 0, 0, 0)
 /* clang-format on */
@@ -4573,7 +4573,7 @@ na_ofi_eq_open(const struct na_ofi_fabric *na_ofi_fabric,
     cq_attr.wait_cond = FI_CQ_COND_NONE;
     cq_attr.format = FI_CQ_FORMAT_TAGGED;
     cq_attr.size = NA_OFI_CQ_DEPTH;
-    if (cpu > 0) {
+    if (cpu >= 0) {
         NA_LOG_SUBSYS_DEBUG(ctx, "Setting CQ signaling_vector to cpu %d", cpu);
         cq_attr.flags = FI_AFFINITY;
         cq_attr.signaling_vector = cpu;
